src/Aspire.Hosting/Exec/ContainerExecService.cs

```diff
@@ -0,0 +1,146 @@
+
+    /// <summary>
+    /// Execute a command for the specified resource.
+    /// </summary>
+    /// <param name="resourceId">The specific id of the resource instance.</param>
+    /// <param name="commandName">The command name.</param>
+    /// <returns>The <see cref="ExecuteCommandResult" /> indicates command success or failure.</returns>
+    public ExecCommandRun ExecuteCommand(string resourceId, string commandName)
+    {
+        if (!_resourceNotificationService.TryGetCurrentState(resourceId, out var resourceEvent))
+        {
+            return new()
+            {
+                ExecuteCommand = token => Task.FromResult(CommandResults.Failure($"Failed to get the resource {resourceId}"))
+            };
+        }
+
+        var resource = resourceEvent.Resource;
+        if (resource is not ContainerResource containerResource)
+        {
+            throw new ArgumentException("Resource is not a container resource.", nameof(resourceId));
+        }
+
+        return ExecuteCommand(containerResource, commandName);
+    }
+
+    public ExecCommandRun ExecuteCommand(ContainerResource containerResource, string commandName)
+    {
+        var annotation = containerResource.Annotations.OfType<ResourceExecCommandAnnotation>().SingleOrDefault(a => a.Name == commandName);
+        if (annotation is null)
+        {
+            return new()
+            {
+                ExecuteCommand = token => Task.FromResult(CommandResults.Failure($"Failed to get the resource {containerResource.Name}"))
+            };
+        }
+
+        return ExecuteCommandCore(containerResource, annotation.Name, annotation.Command, annotation.WorkingDirectory);
+    }
+
+    /// <summary>
+    /// Executes a command for the specified resource.
+    /// </summary>
+    /// <param name="resource">The resource to execute a command in.</param>
+    /// <param name="commandName"></param>
+    /// <param name="command"></param>
+    /// <param name="workingDirectory"></param>
+    /// <returns></returns>
+    private ExecCommandRun ExecuteCommandCore(
+        ContainerResource resource,
+        string commandName,
+        string command,
+        string? workingDirectory)
+    {
+        var resourceId = resource.GetResolvedResourceNames().First();
+
+        var logger = _resourceLoggerService.GetLogger(resourceId);
+        logger.LogInformation("Starting command '{Command}' on resource {ResourceId}", command, resourceId);
+
+        var containerExecResource = new ContainerExecutableResource(commandName, resource, command, workingDirectory);
+        _dcpNameGenerator.EnsureDcpInstancesPopulated(containerExecResource);
+        var dcpResourceName = containerExecResource.GetResolvedResourceName();
+
+        Func<CancellationToken, Task<ExecuteCommandResult>> commandResultTask = async (CancellationToken cancellationToken) =>
+        {
+            await _dcpExecutor.RunEphemeralResourceAsync(containerExecResource, cancellationToken).ConfigureAwait(false);
```

Comment:
It seems awkward that the call to `ContainerExecService.ExecuteCommand()` does not, in fact, executes the command. 

Consider making this method an async method and taking the call to `RunEphemenralResourceAsync()` out of the result-producing task. This will guarantee that when `ExecuteCommandCore()` returns without exception, the command is started. The returned run object would then expose `GetResultAsync()` method for waiting on the result. 

This change also guarantees that calling `GetOutputStream()` on the returned run object always makes sense. Without it, it is possible to ask for output stream without actually starting the command, which will result in waiting forever (until the cancellation token expires)